hashes/zkevm/src/keccak/coprocessor/circuit/leaf.rs

```diff
@@ -0,0 +1,473 @@
+        POSEIDON_R_P,
+        POSEIDON_SECURE_MDS,
+    >();
+    PoseidonHasher::<F, POSEIDON_T, POSEIDON_RATE>::new(spec)
+}
+
+/// Encode raw inputs from Keccak circuit witnesses into lookup keys.
+///
+/// Each element in the return value corrresponds to a Keccak chunk. If is_final = true, this element is the lookup key of the corresponding logical input.
+fn encode_inputs_from_keccak_fs<F: Field>(
+    ctx: &mut Context<F>,
+    range_chip: &impl RangeInstructions<F>,
+    initialized_hasher: &PoseidonHasher<F, POSEIDON_T, POSEIDON_RATE>,
+    loaded_keccak_fs: &[LoadedKeccakF<F>],
+) -> Vec<PoseidonCompactOutput<F>> {
+    // Circuit parameters
+    let num_poseidon_absorb_per_keccak_f = num_poseidon_absorb_per_keccak_f::<F>();
+    let num_word_per_witness = num_word_per_witness::<F>();
+    let num_witness_per_keccak_f = POSEIDON_RATE * num_poseidon_absorb_per_keccak_f;
+
+    // Constant witnesses
+    let rate_witness = ctx.load_constant(F::from(POSEIDON_RATE as u64));
+    let one_witness = ctx.load_constant(F::ONE);
+    let zero_witness = ctx.load_zero();
+    let multiplier_witnesses = ctx.load_constants(&get_words_to_witness_multipliers::<F>());
+
+    let compact_input_len = loaded_keccak_fs.len() * num_poseidon_absorb_per_keccak_f;
+    let mut compact_inputs = Vec::with_capacity(compact_input_len);
+    let mut is_final_last = one_witness;
+    for loaded_keccak_f in loaded_keccak_fs {
+        // If this keccak_f is the last of a logical input.
+        let is_final = loaded_keccak_f.is_final;
+        let mut poseidon_absorb_data = Vec::with_capacity(num_witness_per_keccak_f);
+
+        // First witness of a keccak_f: [<len_word>, word_values[0], word_values[1], ...]
+        // <len_word> is the length of the input if this is the first keccak_f of a logical input. Otherwise 0.
+        let mut words = Vec::with_capacity(num_word_per_witness);
+        let len_word =
+            range_chip.gate().select(ctx, loaded_keccak_f.bytes_left, zero_witness, is_final_last);
+        words.push(len_word);
+        words.extend_from_slice(&loaded_keccak_f.word_values[0..(num_word_per_witness - 1)]);
+        let first_witness = range_chip.gate().inner_product(
+            ctx,
+            multiplier_witnesses.clone(),
+            words.iter().map(|w| halo2_base::QuantumCell::Existing(*w)),
+        );
+        poseidon_absorb_data.push(first_witness);
+
+        // Turn every num_word_per_witness words later into a witness.
+        for words in &loaded_keccak_f
+            .word_values
+            .into_iter()
+            .skip(num_word_per_witness - 1)
+            .chunks(num_word_per_witness)
+        {
+            let mut words = words.collect_vec();
+            words.resize(num_word_per_witness, zero_witness);
+            let witness = range_chip.gate().inner_product(
+                ctx,
+                multiplier_witnesses.clone(),
+                words.iter().map(|w| halo2_base::QuantumCell::Existing(*w)),
```

Comment:
switch the order so `words` is first argument and then `multiplier` so you don't need to do this `Existing` thing. This will lead to an optimization since `multiplier[0] = F::ONE`